batch/batch/server/server.py

```diff
@@ -422,14 +428,46 @@ def to_dict(self):
         return result
 
 
+with open(os.environ.get('HAIL_JWT_SECRET_KEY_FILE') or '/jwt-secret/secret-key') as f:
+    jwtclient = hj.JWTClient(f.read())
+
+
+def authenticated_users_only(fun):
+    def wrapped(request, *args, **kwargs):
+        encoded_token = request.cookies.get('user')
+        if encoded_token is not None:
+            try:
+                userdata = jwtclient.decode(encoded_token)
+                if 'userdata' in fun.__code__.co_varnames:
+                    return fun(request, *args, userdata=userdata, **kwargs)
+                return fun(request, *args, **kwargs)
+            except jwt.exceptions.DecodeError as de:
+                log.info(f'could not decode token: {de}')
+        raise web.HTTPForbidden()
```

Comment:
We’re preferring 404
 (This previously came up in a notebook2 Pr I can dig up)
https://stackoverflow.com/questions/28582830/access-denied-403-or-404

Reply:
I disagree. It seems helpful to me as someone who might debug things to know that `/jobss` is 404 and `/jobs` is 403. We're not leaking the existence or non-existence of secret resources, just the existence or non-existence of an api endpoint.

I agree, in the forthcoming PR that protects individual resources (i.e. `/job/1`) we should return 404 if the user is not authorized to access said resource.

Ideally this would be 401 unauthorized because you need to go authenticate, but 401 has special meaning in HTTP.

Reply:
We don't want it to be helpful, we want it to uniformly deny requests. Anything that is helpful leaks information. We, as the designers of the system know to provide a credential. We don't want someone who randomly pings addresses to get a 403 and decide that it's worth investing more time into.

Reply:
We *should* know, but we make mistakes. We should help ourselves when possible to do so without damaging user security and privacy.

My understanding of the policy of conflating forbidden and missing is: the existence of a resource is one bit of *secret* information. We must not reveal secret information, ergo we must not leak that bit.

A common example is when the name of the resource is meaningful: if `batch.hail.is/jobs/arcturus-birthday-present` returns 403, then Arcturus has found out that we're getting them a birthday present. That's a privacy failure.

Another example is learning the system load. An actor could probe `/jobs/1`, `/jobs/2`, `/jobs/4`, ... to search for how many jobs exist. If we consider the system load private information (and I can accept that system load is private-to-hail-team information), then leaking it is a privacy failure.

Revealing public information is, by definition, not a privacy failure or information leak. I assert (and maybe this is where we disagree) that the batch API itself is public information. That batch will respond to `/jobs` or `/recent` is public information. The information batch responds with might be private, but the fact that batch responds at all to those endpoints, does not seem, to me, to be private information.

This is not an unusual perspective. Spotify and GitHub, for example, both return 401 Unauthorized (though GitHub violates the requirement to include a WWW-Authenticate header), if you attempt to access authentication restricted API endpoints:

```
(hail) dking@wmb16-359 # curl -vsL https://api.spotify.com/v1/me 2>&1 >/dev/null | grep -e '< '    
< HTTP/1.1 401 Unauthorized
< Content-Type: application/json
< WWW-Authenticate: Bearer realm="spotify"
< Content-Type: application/json
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Headers: Accept, Authorization, Origin, Content-Type, Retry-After
< Access-Control-Allow-Methods: GET, POST, OPTIONS, PUT, DELETE, PATCH
< Access-Control-Allow-Credentials: true
< Access-Control-Max-Age: 604800
< Content-Length: 74
< Date: Fri, 12 Apr 2019 18:26:51 GMT
< Via: 1.1 google
< Alt-Svc: clear
< 
(hail) dking@wmb16-359 # curl -vsL https://api.github.com/user 2>&1 >/dev/null | grep -e '< '                        
< HTTP/1.1 401 Unauthorized
< Server: GitHub.com
< Date: Fri, 12 Apr 2019 18:27:27 GMT
< Content-Type: application/json; charset=utf-8
< Content-Length: 135
< Status: 401 Unauthorized
< X-RateLimit-Limit: 60
< X-RateLimit-Remaining: 56
< X-RateLimit-Reset: 1555097247
< X-GitHub-Media-Type: github.v3; format=json
< Access-Control-Expose-Headers: ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type
< Access-Control-Allow-Origin: *
< Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
< X-Frame-Options: deny
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< Referrer-Policy: origin-when-cross-origin, strict-origin-when-cross-origin
< Content-Security-Policy: default-src 'none'
< X-GitHub-Request-Id: F550:4536:5EB0E4:BE85B5:5CB0D88F
< 
```

I think the right answer is clearly to return 401 Unauthorized with `WWW-Authenticate` set to `Bearer`. I was unaware of 401 when I wrote this, it is clearly the more semantically appropriate status code in this case.

I do agree that when we implement individual resources access control in batch (which this PR does not implement), `/jobs/0` should return:
- 401 Unauthorized, if the user presents an invalid token (or no token at all)
- 404 Not Found, if user is not authorized to view that resource or if the resource does not exist

I may even accept always returning 404 Not Found, but I really would rather not. I think an invalid token (e.g. one that has expired) should be distinguishable from lack of authorization or lack of existence.

Reply:
I think that you should return a 403, or 401, if it meaningfully benefits your api. If not, I still think a 404 is appropriate. The less information reveled the better: as you mentioned, do you want foreign agents who don't already know of your endpoint to learn that you serve it? I'd argue that your API is made public through documentation and web links, not through your error code. The people who don't read those shouldn't have an easier time learning of them.

The situation I've found a 401 to be useful is when I need to indicate to the api that it must re-authenticate (expired token), or must log in. If you have this scenario, great. I agree that 403 is semantically incorrect; you could issue this to indicate future login attempts would be denied.